### ISSUE TITLE
Dart 2.5.0-dev.1.0 in AOT mode strange behavior annotation with Function reference

### ISSUE THREAD

**Issue by hpoul** [dart-lang/sdk#37656]
I was debugging a crash which only occurred in release mode in a flutter application on master channel, but not in stable. I was now able to reproduce the problem in AOT mode in dart 2.5, but it works in dart VM jit mode and dart 2.4. 
It seems the AOT compiler does something weird when a property is annotated with a Function reference (even if this function reference is never accessed).

I have reduced the problem to the following code:

```dart
class SomeTestAnnotation {
  const SomeTestAnnotation({this.fromJson});
  final Function fromJson;
}

class SomeClass {
  SomeClass() : someValue = 'initial value' {}

  @SomeTestAnnotation(fromJson: idConvert)
  String someValue;

  static String idConvert(dynamic jsonValue) => '';
}

void main() {
  print('creating instance.');
  final a = SomeClass();
  print('a: ${a.someValue}');
  a.someValue = 'updated value';
  print('a: ${a.someValue}');
  print('finished');
}
```

expected behavior in dart VM:
```bash
Herbys-MacBook-Pro-2017:lib$ dart another_try.dart
creating instance.
a: initial value
a: updated value
finished
```

but when using AOT the value of the property is still null:

```
Herbys-MacBook-Pro-2017:lib$ dart2aot another_try.dart another_try.dart.aot
Herbys-MacBook-Pro-2017:lib$ dartaotruntime another_try.dart.aot
creating instance.
a: null
a: updated value
finished
```

The problem goes away when either:
* Remove the annotation completely
* just the function reference
* use a top level function reference like `@SomeTestAnnotation(fromJson: print)`

Because of some reason the value used during initialization is not assigned to the property? When this is used inside a more complicated code (I have used the json_serializable package to generate json deserialization code) it even crashes the AOT runtime with a segfault:

```
===== CRASH =====
si_signo=Segmentation fault: 11(11), si_code=1, si_addr=0xbb
Abort trap: 6
```

i am using mac os with the latest dart 2.5 dev version from homebrew (but since i stumbled onto this crash also on iOS and Android, i'm pretty sure it's not limited to the mac build):
```
Herbys-MacBook-Pro-2017:lib$ dart --version
Dart VM version: 2.5.0-dev.1.0 (Tue Jul 9 15:27:01 2019 +0200) on "macos_x64"
```

I haven't found any issue tracking this problem, and i realise this is still in development, so feel free to close it if it's redundant.

**Comment by mraleph** [comment]
@alexmarkov does this ring a bell? This does not reproduce on master HEAD - so it might be fixed.

**Comment by megadrazz** [comment]
e muito relato de problemas esse dart isso ta atrasando a vida de qualquer um!!! aqui da error de mais mano vou migra pra outro fui..

**Comment by alexmarkov** [comment]
Bisected this bug: the problem appeared since b32d196f00debeb38300d1a89c120958ceeda214, and was fixed in 63120303a7496cb36967b69937cadd2c39e82386.

It's likely a duplicate to https://github.com/dart-lang/sdk/issues/37544. The fix 63120303a7496cb36967b69937cadd2c39e82386 added a check for recursive finalization when reading fields; without that check compiler could see fresh non-finalized fields with invalid offsets (if annotation on a field caused recursive finalization of the current class).
